packages/astro/src/core/build/plugins/plugin-pages.ts

```diff
@@ -1,54 +1,76 @@
 import type { Plugin as VitePlugin } from 'vite';
-import { pagesVirtualModuleId, resolvedPagesVirtualModuleId } from '../../app/index.js';
 import { addRollupInput } from '../add-rollup-input.js';
-import { eachPageData, type BuildInternals } from '../internal.js';
+import { type BuildInternals } from '../internal.js';
 import type { AstroBuildPlugin } from '../plugin';
 import type { StaticBuildOptions } from '../types';
 import { MIDDLEWARE_MODULE_ID } from './plugin-middleware.js';
 import { RENDERERS_MODULE_ID } from './plugin-renderers.js';
+import { extname } from 'node:path';
+
+export const ASTRO_PAGE_MODULE_ID = '@astro-page:';
+export const ASTRO_PAGE_RESOLVED_MODULE_ID = '\0@astro-page:';
+
+// This is an arbitrary string that we are going to replace the dot of the extension
+export const ASTRO_PAGE_EXTENSION_POST_PATTERN = '@_@';
 
 function vitePluginPages(opts: StaticBuildOptions, internals: BuildInternals): VitePlugin {
 	return {
 		name: '@astro/plugin-build-pages',
 
 		options(options) {
 			if (opts.settings.config.output === 'static') {
-				return addRollupInput(options, [pagesVirtualModuleId]);
+				const inputs: Set<string> = new Set();
+
+				for (const path of Object.keys(opts.allPages)) {
+					const extension = extname(path);
+
+					// we mask the extension, so this virtual file
+					// so rollup won't trigger other plugins in the process
+					const virtualModuleName = `${ASTRO_PAGE_MODULE_ID}${path.replace(
+						extension,
+						extension.replace('.', ASTRO_PAGE_EXTENSION_POST_PATTERN)
+					)}`;
+
+					inputs.add(virtualModuleName);
+				}
+
+				return addRollupInput(options, Array.from(inputs));
 			}
 		},
 
 		resolveId(id) {
-			if (id === pagesVirtualModuleId) {
-				return resolvedPagesVirtualModuleId;
+			if (id.startsWith(ASTRO_PAGE_MODULE_ID)) {
+				return '\0' + id;
 			}
 		},
 
 		async load(id) {
-			if (id === resolvedPagesVirtualModuleId) {
-				let importMap = '';
+			if (id.startsWith(ASTRO_PAGE_RESOLVED_MODULE_ID)) {
 				const imports: string[] = [];
 				const exports: string[] = [];
-				const content: string[] = [];
-				let i = 0;
-				imports.push(`import { renderers } from "${RENDERERS_MODULE_ID}";`);
-				exports.push(`export { renderers };`);
-				for (const pageData of eachPageData(internals)) {
-					const variable = `_page${i}`;
-					imports.push(
-						`const ${variable} = () => import(${JSON.stringify(pageData.moduleSpecifier)});`
-					);
-					importMap += `[${JSON.stringify(pageData.component)}, ${variable}],`;
-					i++;
-				}
+				// split by ":", the second element is the page name, which will start with "src/..."
+				const [, pageName] = id.split(':');
+				// We replaced the `.` of the extension with ASTRO_PAGE_EXTENSION_POST_PATTERN, let's replace it back
```

Comment:
nit: usually the pattern with handling virtual files is using `id.slice(ASTRO_PAGE_RESOLVED_MODULE_ID.length)`, which I think would be safer too.